### DISCUSSION TITLE
Project and package naming

### DISCUSSION THREAD

**Discussion by woodseowl** [CornellCustomDev/CD-LaravelStarterKit#20] [General]
It seems we need to come to a consensus on naming soon, based on some discussion in Slack from Nov. 18:

Rich:
> I see you used the namespace CUCustomDev.  Are we settled on that?  I had used CUCustomApps for my feed project, but can switch that.

Alison:
> I'm surprised you'd use capital letters in the project name?

Irina:
> I agree about the name

Eric:
> My comments from PR #16:
> - The composer.json name I used for the package is cu-communityapps/cd-laravelstarterkit since it's the CU-CommunityApps/CD-LaravelStarterKit github repo.
> - The PSR-4 namespace I used is CUCustomDev\LaravelStarterKit, which is mapped in the composer file. This doesn't match the repo name because I'm trying to figure out where we are going to put other Laravel packages for custom dev. I hear we are getting our own GitHub org, which might make naming like this work. I'm looking for feedback.
>
> I started down these naming paths simply based on recent projects in our GitHub org. I agree that this is not in keeping with standard naming in the git repo world at large. This also will intersect with [packagist.org](http://packagist.org/) ownership and naming questions. I'd love to have someone with some opinions and authority on this set us on a good path.

Perhaps we can continue the discussion here and settle on a couple rules for naming?

**Comment by inaydich** [comment]
If we keep "CUCustomDev\LaravelStarterKit", will the package name be just "LaravelStarterKit"? I think we should include CD in there somehow.

**Comment by richmarisa** [comment]
As a proposed example the repository would be github.com/CUCustomDev/LaravelStarterKit; the name package would be cucustomdev/laravel-starter-kit, and the namespace for the package would be \CUCustomDev\LaravelStarterKit containing classes like \CUCustomDev\LaravelStarterKit::WhateverClass

We just included CD in our names because we were sharing the top level repository name.  If the top name is CUCustomDev, then I think the CD would be redundant.

**Comment by woodseowl** [comment]
I'd like to see consistent use of capitals and dashes. It looks to me like "cu-custom-dev/laravel-starter-kit" would be more consistent than "cucustomdev/laravel-starter-kit" for the package name. 

I'm still personally a bit uneasy with the inconsistency in the repo name vs package name, thinking kebab-case is the more typical approach in the wild. But I don't have a strong opinion or best practice guidance to inform that.

I agree that having CD in the package repository name is redundant once we are in our own CU Custom Dev repo.

**Reply by richmarisa** [reply]
I was just following the pattern of other packages in composer.json.

**Reply by richmarisa** [reply]
Ok, I found examples of sashes in the first part of the name. I don't object to that.

**Reply by richmarisa** [reply]
*dashes

**Comment by woodseowl** [comment]
In our Devs meeting on Dec. 2, we discussed naming conventions for namespaces and packages. It was boisterous and opinionated! There was not a clear consensus on a single approach, but here were things that people valued:

- Have "Cornell" in the name rather than "CU" in order to reduce potential confusion or namespace collision (e.g., entities like University of Colorado and Columbia University)
- Have the name be identifiable as related to our team
- Have the name be as short as practical
- Have some fun/whimsy like our "cubear" packagist vendor name

When actually putting those values into practice, there were three names that settled out as potentials, but no clear winners (and these would be potential names for a GitHub organization just for Custom Development):

- CornellCD (trying to be short)
- CornellCustomDev (trying to be easily identifiable)
- CornellBear (picking up from the packagist "cubear" naming)

For packages or custom modules, the composer package version of that naming would end up being in the form "repo-name/package-name", like:

- cornell-cd/package-name
- cornell-custom-dev/package-name
- cornell-bear/module-name

In the Devs meeting, the suggestion was to let people in Custom Dev do a ranked-choice voting as a way to choose one of these names.

**Reply by inaydich** [reply]
looks good

**Comment by woodseowl** [comment]
Voting results: https://app.rankedvote.co/decisions/27623/Custom-Dev-namespace-naming/28744/results

- 6 people put CornellCustomDev first, 5 had CornellCD first, and 3 had CornellBear first (no majority winner)
- Of the CornellBear first voters, 2 had CornellCustomDev second, and 1 had CornellCD second
- Final count was 8 for CornellCustomDev and 6 for CornellCD (majority winner)

**Comment by woodseowl** [comment]
Based on the voting results, I suggest a packagist name of cornell-custom-dev/laravel-starter-kit and PHP code namespace of /CornellCustomDev/LaravelStarterKit. That follows prevailing name formatting that I see in packagist.
